### PR TITLE
Redo Exhaustiveness Analysis

### PR DESCRIPTION
**Pull request by CodaFi**
Implement exhaustiveness checking in Sema with rich error messages.  The
algorithm used is a variant of the one described in Fengyun Liu's paper
"A Generic Algorithm for Checking Exhaustivity of Pattern Matching"
published in the EPFL conference, and Luc Maranget's seminal paper
"Warnings for Pattern Matching"

The Space Engine views pattern matching as a problem of projecting the
scrutinee of a pattern-match into a "Space", then iteratively
constructing a Space from the cases.  Taking the difference of this
master space and the covered spaces yields the "holes" left over or
reveals a completely covered space.

The algorithm also extends trivially to redundancy checks in patterns,
but that check is already implemented in SIL dataflow diagnostics and 
it's not clear if this this algorithm improves upon it.

Resolves [SR-483](https://bugs.swift.org/browse/SR-483), [SR-2293](https://bugs.swift.org/browse/SR-2293), [SR-1313](https://bugs.swift.org/browse/SR-1313), [SR-4468](https://bugs.swift.org/browse/SR-4468), [SR-911](https://bugs.swift.org/browse/SR-911), [SR-4658](https://bugs.swift.org/browse/SR-4658), [SR-3768](https://bugs.swift.org/browse/SR-3768) and probably a few others.

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by jrose-apple** [comment]
cc @gottesmm

**Comment by CodaFi** [comment]
@swift-ci please smoke test macOS platform

**Comment by CodaFi** [comment]
@swift-ci please smoke test macOS platform

**Comment by CodaFi** [comment]
@swift-ci please smoke test macOS platform

**Comment by CodaFi** [comment]
@swift-ci please smoke test macOS platform

**Comment by CodaFi** [comment]
@jrose-apple I'm baffled about the final failing validation test.  Inserting defaults causes the driver not to skip building `main`.

**Comment by CodaFi** [comment]
@swift-ci please smoke test Linux platform

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeCheckStmt.cpp [hunk 1]
```diff
@@ -377,6 +306,1039 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
   /// expressions are not discarded.
   bool IsREPL;
 
+  /// The SpaceEngine encapsulates an algorithm for computing the exhaustiveness
```
Comment (slavapestov) [inline lib/Sema/TypeCheckStmt.cpp]: How about putting this in its own file?

File: test/ClangImporter/ctypes_parse.swift [hunk 2]
```diff
@@ -9,6 +9,7 @@ func checkEquatablePattern(_ c: Color) {
     case red: return
     case green: return
     case blue: return
+    default: return
```
Comment (slavapestov) [inline test/ClangImporter/ctypes_parse.swift]: Is this a source-breaking change?

**Comment by slavapestov** [comment]
I don't know much about this problem domain, but resolving five JIRA bugs is impressive ;)

File: test/ClangImporter/ctypes_parse.swift (hunk 2 above)
Comment (CodaFi) [inline test/ClangImporter/ctypes_parse.swift]: I think so, but at the same time it's also something that the old check wasn't catching but should have - and there's a trivial migration path here.

File: lib/Sema/TypeCheckStmt.cpp (hunk 1 above)
Comment (CodaFi) [inline lib/Sema/TypeCheckStmt.cpp]: TypeCheckSwitchStmt.cpp?

File: lib/Sema/TypeCheckStmt.cpp [hunk 3]
```diff
@@ -377,6 +306,1039 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
...
+        if (tp->isBool()) {
+          arr.push_back(Space(true));
+          arr.push_back(Space(false));
+        } else if (auto *E = tp->getEnumOrBoundGenericEnum()) {
+          // Look into each case of the enum and decompose it in turn.
+          auto children = E->getAllElements();
+          std::transform(children.begin(), children.end(),
+                         std::back_inserter(arr), [&](EnumElementDecl *eed) {
+            // FIXME: This shouldn't happen.
+            if (!eed->hasInterfaceType()) {
```
Comment (CodaFi) [inline lib/Sema/TypeCheckStmt.cpp]: @slavapestov Can you explain this?  There's one test case in our suite that causes an enum element decl to be created sans interface type but still pass typechecking.

Comment (slavapestov) [inline lib/Sema/TypeCheckStmt.cpp]: Which test?

**Comment by jckarter** [comment]
Thanks for looking into this! I have concerns about doing the checking in Sema though, since that means we end up with duplicated logic for understanding pattern exclusivity, reachability, and exhaustiveness across Sema, SILGen, and the mandatory passes. I had envisioned tackling this problem by improving unreachable code diagnostics' understanding of switch_enums.

**Comment by gottesmm** [comment]
I agree 100% with @jckarter.

**Comment by jckarter** [comment]
I'm willing to believe sema is a better home for this analysis, but it might do some good to tease out the specialization, destructuring, and other abstractions from SILGenPattern into Sema so they can be shared.

**Comment by CodaFi** [comment]
> since that means we end up with duplicated logic

I was hoping to be able to remove SIL's ability to reason about pattern completeness entirely after a while.  This check in Sema is both more conservative and more complete - the old check would recur into tuple patterns and yield false positives, for example, but leave the dataflow checks in place as a way to diagnose redundancy and as a sanity check to back up the exhaustiveness checker.

**Comment by jckarter** [comment]
Aside from exhaustiveness checking, it seems to me like there's other duplicated logic shared between the exhaustiveness analysis of patterns and code generation. In both cases, you need an understanding of what patterns are mutually exclusive and how to specialize the pattern graph after applying a pattern. I'm wondering whether there are abstractions your space engine can share with the SILGenPattern algorithm, so we don't have to duplicate the work every time we introduce a new kind of pattern.

**Comment by CodaFi** [comment]
In both of the cited papers the algorithms don't focus on decision-flows, they're just syntactico-semantic checks for exhaustiveness.  Given the precondition that a pattern is complete, one can use this approach to detect overlap in patterns - and I did originally plan to do that - but SIL does it far better than this currently.  Compiling patterns to decision trees, unfortunately, seems like a different problem.

**Comment by CodaFi** [comment]
@swift-ci please smoke test macOS platform

**Comment by CodaFi** [comment]
@swift-ci please smoke test macOS platform

File: lib/Sema/TypeCheckStmt.cpp (hunk 3 above)
Comment (CodaFi) [inline lib/Sema/TypeCheckStmt.cpp]: If you replace this return with an `llvm_unreachable`, then `test/NameBinding/reference-dependencies.swift` hits it while type-checking `lookUpManyTopLevelNames`.

**Comment by CodaFi** [comment]
@swift-ci please smoke test macOS platform

**Comment by CodaFi** [comment]
@swift-ci please smoke test macOS platform

**Comment by CodaFi** [comment]
@swift-ci please smoke test Linux platform

**Comment by CodaFi** [comment]
@swift-ci please smoke test Linux platform

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
Alright: @jckarter @gottesmm I don't want to merge this without convincing y'all that this is the approach we should take for this feature.  

I've poked around in our implementation of Maranget's algorithm for minimal decision tree generation for the last few days and I can't really see it fitting in with the idea of specialization of trees.  For one, the iterative specialization process doesn't maintain the shape of the case matrix as the user wrote it.  Even if the process were perfect, if we attempted to diagnose non-exhaustive specialized rows we may wind up introducing far more patterns than needed.  The benefit of this approach is it arrives at a minimal set of uncovered cases fairly efficiently and can emit diagnostics that match the actual structure the user wrote.

**Comment by slavapestov** [comment]
I think having this check in Sema makes sense. I don't really see how exhaustiveness checking benefits from having a CFG around. @gottesmm @jckarter what do you think?

**Comment by CodaFi** [comment]
Ping @jckarter.  I'd really like to land this in Swift 4 too

**Comment by CodaFi** [comment]
Let's see what happens here too

@swift-ci please smoke source compatibility.

**Comment by CodaFi** [comment]
Let's see what happens here too.

@swift-ci please test source compatibility

**Comment by CodaFi** [comment]
Apparently GRDB is still acting up @lplarson

**Comment by lplarson** [comment]
Hmm, not seeing that in CI.

**Comment by lplarson** [comment]
Let's try this one more time.

@swift-ci please test source compatibility

**Comment by lplarson** [comment]
This looks like real source breakage:
https://ci.swift.org/job/swift-PR-source-compat-suite/13/artifact/swift-source-compat-suite/FAIL_BuildSwiftPackage_GRDB-swift.log

**Comment by CodaFi** [comment]
Ah, you're right.  Didn't get the semantic pattern of a paren pattern.

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please clean smoke test macOS platform

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility

**Comment by CodaFi** [comment]
After a short conversation with Joe, I feel comfortable with this going in.

**Comment by lattner** [comment]
Awesome @CodaFi !

**Comment by jckarter** [comment]
@CodaFi There's a validation test that raises exhaustiveness errors in Sema now:

```
validation-test/Driver/Dependencies/Inputs/rdar25405605/helper-2.swift:8:5: error: switch must be exhaustive
    switch value {
    ^
validation-test/Driver/Dependencies/Inputs/rdar25405605/helper-2.swift:8:5: note: add missing case: '.three'
    switch value {
    ^
```

Mind updating the test?

**Pull request merged**